extensions/amp-auto-lightbox/0.1/carousel-criteria.js

```diff
@@ -0,0 +1,294 @@
+  /**
+   * @param {!Element} element
+   * @return {!Promise<boolean>}
+   */
+  static meetsAll(element) {
+    const win = toWin(element.ownerDocument.defaultView);
+
+    if (!isExperimentOn(win, 'amp-auto-lightbox-carousel')) {
+      return resolveFalse();
+    }
+
+    const slides = element.querySelectorAll('.amp-carousel-slide');
+    const images = element.querySelectorAll('amp-img');
+
+    if (images.length < 1) {
+      return resolveFalse();
+    }
+
+    if (slides.length != images.length) {
+      return resolveFalse();
+    }
+
+    let promise = resolveTrue();
+
+    iterateCursor(slides, slide => {
+      promise = promise.then(previousWasAccepted => {
+        if (!previousWasAccepted) {
+          return false;
+        }
+        return SlideCriteria.meetsAll(slide);
+      });
+    });
+
+    return promise;
+  }
+}
+
+class SlideCriteria {
+
+  /**
+   * @param {!Element} element
+   * @return {!Promise<boolean>}
+   */
+  static meetsAll(element) {
+    if (element.tagName == 'AMP-IMG') {
+      // Already traversed up the tree, we only need to check the immediate
+      // action.
+      const actionLookupStopAt = dev().assertElement(element.parentElement);
+      const actions = Services.actionServiceForDoc(element);
+      return tryResolve(() =>
+        !actions.hasResolvableAction(element, 'tap', actionLookupStopAt));
+    }
+
+    const img = element.querySelector('amp-img');
+    if (!img) {
+      return resolveFalse();
+    }
+
+    const slideMeetsSizingPromise =
+        SlideCriteria.meetsSizingCriteria(img, element);
+
+    return slideMeetsSizingPromise.then(slideMeetsSizingCriteria => {
+      if (!slideMeetsSizingCriteria) {
+        return false;
+      }
+      return SlideCriteria.meetsAllSync(element, /* depth */ 0);
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {number} depth
+   * @return {boolean}
+   */
+  static meetsAllSync(element, depth) {
+    if (depth > MAX_TRAVERSING_DEPTH) {
+      return false;
+    }
+    if (!isElementAdmissible(element)) {
+      return false;
+    }
+    // Already traversed up the tree, we only need to check the immediate
+    // action.
+    const actionLookupStopAt = dev().assertElement(element.parentElement);
+    const actions = Services.actionServiceForDoc(element);
+    if (actions.hasResolvableAction(element, 'tap', actionLookupStopAt)) {
+      return false;
+    }
+    if (!shouldElementBeTraversed(element)) {
+      return true;
+    }
+    const {children} = element;
+    if (children.length > MAX_TRAVERSING_BREADTH) {
+      return false;
+    }
+    return toArray(children).every(child =>
+      SlideCriteria.meetsAllSync(child, depth + 1));
+  }
+
+  /**
+   * @param {!AmpElement} img
+   * @param {!Element} slide
+   * @return {!Promise<boolean>}
+   */
+  static meetsSizingCriteria(img, slide) {
+    devAssert(img.tagName == 'AMP-IMG');
+
+    return img.getImpl().then(impl => new Promise(resolve => {
```

Comment:
I believe `measureElement` returns a `Promise` that contains the return value of the `measurer` function, so this new wrapper `Promise` is redundant. You can do

```js
return img.getImpl().then(impl => {
  return impl.measureElement(() => {
    /*...*/
    return (imageArea/slideArea >= MIN_IMG_SLIDE_AREA_RATIO);
  });
});

Reply:
Unfortunately the measure promise is shared by the entire batch so returning values is not possible.